VariantShredding.md

```diff
@@ -25,290 +25,316 @@
+Readers can assume that a value is not an object if `typed_value` is null and that `typed_value` field values are correct; that is, readers do not need to read the `value` column if `typed_value` fields satisfy the required fields.
 
-```
-[
-  {
-    "a": [1, 2, 3],
-    "b": 100
-  },
-  {
-    "a": [4, 5, 6],
-    "b": 200
-  }
-]
-```
+## Nesting
 
+The `typed_value` associated with any Variant `value` field can be any shredded type, as shown in the sections above.
 
-The corresponding Parquet schema with “a” and “b” as leaf types is:
+For example, the `event` object above may also shred sub-fields as object (`location`) or array (`tags`).
 
 ```
-optional group variant_col {
- required binary metadata;
- optional binary variant_value;
- optional group array (LIST) {
-  repeated group list {
-   optional group element {
-    optional binary variant_value;
-    optional group object {
-     optional group a {
-      optional binary variant_value;
-      optional group array (LIST) {
-       repeated group list {
-        optional group element {
-         optional int64 typed_value;
-         optional binary variant_value;
+optional group event (VARIANT) {
+  required binary metadata;
+  optional binary value;
+  optional group typed_value {
+    required group event_type {
+      optional binary value;
+      optional binary typed_value (STRING);
+    }
+    required group event_ts {
+      optional binary value;
+      optional int64 typed_value (TIMESTAMP(true, MICROS));
+    }
+    required group location {
+      optional binary value;
+      optional group typed_value {
+        required group latitude {
+          optional binary value;
+          optional double typed_value;
+        }
+        required group longitude {
+          optional binary value;
+          optional double typed_value;
+        }
+      }
+    }
+    required group tags {
+      optional binary value;
+      optional group typed_value (LIST) {
+        repeated group list {
+          required group element {
+            optional binary value;
+            optional binary typed_value (STRING);
+          }
         }
-       }
       }
-     }
-     optional group b {
-      optional int64 typed_value;
-      optional binary variant_value;
-     }
     }
-   }
   }
- }
 }
 ```
 
-In the above example schema, if “a” is an array containing a mix of integer and non-integer values, the engine will shred individual elements appropriately into either `typed_value` or `variant_value`.
-If the top-level Variant is not an array (for example, an object), the engine cannot shred the value and it will store it in the top-level `variant_value`.
-Similarly, if "a" is not an array, it will be stored in the `variant_value` under "a".
+# Data Skipping
 
-Consider the following example:
+Statistics for `typed_value` columns can be used for file, row group, or page skipping when `value` is always null (missing).
```

Comment:
This seems limiting?  If have a predicate like `ExtractElement($.maybe_str_value) = TIMESTAMP_NOW(), couldn't an engine use statistics if the there is a shredded timestamp column (i.e. by definition anything stored in value would not meet this predicate)?

Reply:
Maybe this is related to another question.  Are writers required to shred all values into typed_value that they can?

Reply:
This was careful not to limit the cases where you can infer the statistics can be used. That's why it says **can be used** and **when `value` is always null**. When `value` is null/missing, then you know that all of the values have been shredded and the stats for `typed_value` are correct.

Note that there are other cases where stats are still usable. For instance, if the `value` column's lower and upper bounds are both `00` (Variant-encoded null) then you can trust the stats for the non-null values.

> Are writers required to shred all values into typed_value that they can?

No, but it would limit the ability to take advantage of shredding. There's no value in requiring shredding for all values that can be, because there are still cases where values cannot be shredded (i.e. finding `"n/a"` in an integer column) and we need to be able to handle those non-shredded cases.

Reply:
> There's no value in requiring shredding for all values that can be, because there are still cases where values cannot be shredded (i.e. finding "n/a" in an integer column) and we need to be able to handle those non-shredded cases.

I think this goes back to the bit-ordering problem.  I think the "n/a" case could be identified if you know that all non-shredded values are strings.

Some alternatives here that could still be helpful:

 is to require writers of typed_values to compute stats across all relevant types (even those that are not shredded):
- Exact numeric and floating point (for values that would overflow on conversion, the corresponding statistic isn't set.  For integer<->decimal<->floating point using the rounded extrema would make the statistics useful)
- DateTime with timezone (nanos/micros)
- etc.

This means predicates like `AS_REAL(EXTRACT(col1, '$.a')) > 0.5` could always use stats to prune if the shredded value was any numeric type.

Reply:
I don't think that we can't expect writers to produce custom stats for types based on logical type annotation. If you want stats, then the field should be shredded.

We also can't assume that the semantics of the predicates are to match a type and extract. Those semantics are up to the engine. That's one reason that we need to leave this open-ended.

Reply:
Discussed on dev list and seems like taking a simpler approach for now did not raise objections.